pkg/sql/sem/builtins/aggregate_builtins.go

```diff
@@ -72,19 +72,20 @@ func initAggregateBuiltins() {
 // execution.
 // Exported for use in documentation.
 var Aggregates = map[string][]tree.Builtin{
-	"array_agg": {
-		makeAggBuiltinWithReturnType(
-			[]types.T{types.Any},
+	"array_agg": arrayBuiltin(func(t types.T) tree.Builtin {
+		return makeAggBuiltinWithReturnType(
+			[]types.T{t},
 			func(args []tree.TypedExpr) types.T {
 				if len(args) == 0 {
-					return tree.UnknownReturnType
+					return types.TArray{Typ: t}
 				}
+				// Whenever possible, use the expression's type, so we can properly
+				// handle aliased types that don't explicitly have overloads.
 				return types.TArray{Typ: args[0].ResolvedType()}
```

Comment:
A concern I have with this now is that it effectively creates undocumented overloads for `array_agg` for types that are aliased to types with defined overloads. It might not be a big issue since this function's behaviour is fairly straightforward, but I wanted to at least mention it.